cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -673,72 +684,153 @@ bool solveEquationsSystemWithCholesky(FPType * a, FPType * b, size_t n, size_t n
 }
 
 template <typename FPType, CpuType cpu>
-bool solveEquationsSystemWithPLU(FPType * a, FPType * b, size_t n, size_t nX, bool sequential, bool extendFromSymmetric)
+bool solveEquationsSystemWithSpectralDecomposition(FPType * a, FPType * b, size_t n, size_t nX, bool sequential)
 {
-    if (extendFromSymmetric)
+    /* Storage for the eigenvalues.
+    Note: this allocates more size than they might require when nX > 1, because the same
+    buffer will get reused later on and needs the extra size. Those additional entries
+    will not be filled with eigenvalues. */
+    TArrayScalable<FPType, cpu> eigenvalues(n * nX);
+    DAAL_CHECK_MALLOC(eigenvalues.get());
+
+    /* SYEV parameters */
+    char jobz = 'V';
+    char uplo = 'U';
+    DAAL_INT info;
+
+    /* Query the procedure for size of required buffer */
+    DAAL_INT lwork = -1;
+    FPType buffer_size;
+    if (sequential)
     {
-        /* Extend symmetric matrix to generic through filling of upper triangle */
-        for (size_t i = 0; i < n; ++i)
-        {
-            for (size_t j = 0; j < i; ++j)
-            {
-                a[j * n + i] = a[i * n + j];
-            }
-        }
+        LapackInst<FPType, cpu>::xxsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), &buffer_size, &lwork, &info);
     }
 
-    /* GETRF and GETRS parameters */
-    char trans    = 'N';
-    DAAL_INT info = 0;
+    else
+    {
+        LapackInst<FPType, cpu>::xsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), &buffer_size, &lwork, &info);
+    }
 
-    TArrayScalable<DAAL_INT, cpu> ipiv(n);
-    DAAL_CHECK_MALLOC(ipiv.get());
+    if (info) return false;
 
-    /* Perform P*L*U factorization of A */
+    /* Check that buffer size will not overflow when passed to LAPACK */
+    if (static_cast<size_t>(buffer_size) > std::numeric_limits<DAAL_INT>::max()) return false;
+
+    /* Allocate work buffer as needed */
+    DAAL_INT work_buffer_size = static_cast<DAAL_INT>(buffer_size);
+    TArrayScalable<FPType, cpu> work_buffer(work_buffer_size);
+    DAAL_CHECK_MALLOC(work_buffer.get());
+
+    /* Perform Q*diag(l)*Q' factorization of A */
     if (sequential)
     {
-        LapackInst<FPType, cpu>::xxgetrf((DAAL_INT *)&n, (DAAL_INT *)&n, a, (DAAL_INT *)&n, ipiv.get(), &info);
+        LapackInst<FPType, cpu>::xxsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), work_buffer.get(), &work_buffer_size,
+                                        &info);
     }
     else
     {
-        LapackInst<FPType, cpu>::xgetrf((DAAL_INT *)&n, (DAAL_INT *)&n, a, (DAAL_INT *)&n, ipiv.get(), &info);
+        LapackInst<FPType, cpu>::xsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), work_buffer.get(), &work_buffer_size,
+                                       &info);
+    }
+    if (info) return false;
```

Comment:
Yes, `gesvd` on $\mathbf{X}$ would be more accurate than `syev` on $\mathbf{X}^T \mathbf{X}$, but it would be slower. However the solver `normEq` is based on $\mathbf{X}^T \mathbf{X}$ (I guess that's what the name implies but not 100% sure). `gesvd` could potentially be a fallback from QR, but that'd be a different PR.

I think `scikit-learn-intelex` only calls `oneDAL`'s solver with `normEq` BTW.

Reply:
I meant not to apply `gesvd` to $X$, but rather to $X^T X$.
Due to the factors like rounding errors $X^T X$ might be not positive-semidefinite and `syev` might fail to converge. And we will come to the same issue as we have now with Cholesky.
With SVD we can do:
$X^T X \beta = X^T y$
$X^T X = U \Sigma V^T$
$\beta = V \Sigma^{-1} U^T X^T y$

And there will should be no possibilities to fail.
Or maybe this is too precautious.

Reply:
Yes, it seems that `sklearnex` only uses normal equations method now. We have QR method implemented, but on CPUs only.
And it is rather inefficient in terms of performance to use normal equations at first, and if fail then apply QR to the whole $X$ matrix -> double cost in terms of performance.

Reply:
I don't follow.

The Eigendecomposition part doesn't require positiveness - it would succeed even if the input had negative entries in the main diagonal. If by numerical inaccuracies it happens that some eigenvalue is close to zero in theory but has negative sign when calculated in practice (negative-indefinite matrices would have negative eigenvalues), it would anyway be removed as part of the procedure, so the end result would not take that component.

These "sy" routines takes the values only from the lower triangle of the input matrix, so there shouldn't be any issues with symmetry either if the matrix is getting filled in both corners. I understand they should also be more accurate than `gesvd` when the input matrix is actually symmetric, but I'm not 100% sure about it.

About the QR part: sorry, to clarify - what I meant was that `gesvd` on $\\mathbf{X}$ could be a fallback when `geqrf` on $\\mathbf{X}$ fails. In those cases, $\\mathbf{X}^T \\mathbf{X}$ would not get computed.

Reply:
Yes, right. Forgot that positiveness is not a problem for eigenvalue decomposition, only for Cholesky.
Sorry for misleading info.
My comment was based on the experience with eigenvalue decomposition methods. But I forgot what exactly led to failures. So, I mean those methods can fail and trying to avoid that.

Probably it was not positiveness, but numerical stability issues. 
If the condition number of $X^T X$ is too small, i.e. all the eigenvalues are close, some eigenvalue decomposition methods can fail. Similar things might happen when the condition number is too big.

But I am not too familiar in terms of which MKL decomposition routine is better in the case of small or big condition numbers [thought `gesvd` is the best, but it looks like it's not].

Anyway from your comment I see that for symmetric matrices `sy` methods should be better than `ge`. So, the choice of `syevr` looks reasonable now.

I also understand the point about QR and `gesvd` for $X$ decomposition, but this is not the scope of this PR.

Reply:
Got it. It looks from the [syevd docs](https://www.netlib.org/lapack/explore-3.1.1-html/dsyevd.f.html) that it is indeed possible for this routine to fail due to values in the inputs, although it doesn't explain when that could happen.

That being said, in case it makes it look less bad, if Eigendecomposition fails, it follows that other methods like SciPy's `lstsq` or scikit-learn's `LinearRegression` would also fail since they are also based on spectral decomposition.